backends/ebpf/ebpfTable.h

```diff
@@ -38,44 +58,84 @@ class EBPFTableBase : public EBPFObject {
-        valueTypeName = program->refMap->newName(instanceName + "_value");
+        keyTypeName = instanceName + "_key";
+        valueTypeName = instanceName + "_value";
         dataMapName = instanceName;
     }
 };
 
-class EBPFTable final : public EBPFTableBase {
+class EBPFTable : public EBPFTableBase {
+    const int prefixLenFieldWidth = 32;
+
+    void initKey();
+
+ protected:
+    const cstring prefixFieldName = "prefixlen";
+
+    bool isLPMTable();
+    virtual void validateKeys() const;
+    virtual ActionTranslationVisitor*
+    createActionTranslationVisitor(cstring valueName, const EBPFProgram* program) const {
+        return new ActionTranslationVisitor(valueName, program);
+    }
+
  public:
     const IR::Key*            keyGenerator;
     const IR::ActionList*     actionList;
-    const IR::TableBlock*    table;
+    const IR::TableBlock*     table;
     cstring               defaultActionMapName;
     cstring               actionEnumName;
     std::map<const IR::KeyElement*, cstring> keyFieldNames;
     std::map<const IR::KeyElement*, EBPFType*> keyTypes;
 
     EBPFTable(const EBPFProgram* program, const IR::TableBlock* table, CodeGenInspector* codeGen);
-    void emitTypes(CodeBuilder* builder);
-    void emitInstance(CodeBuilder* builder);
+
+    cstring p4ActionToActionIDName(const IR::P4Action * action) const;
     void emitActionArguments(CodeBuilder* builder, const IR::P4Action* action, cstring name);
-    void emitKeyType(CodeBuilder* builder);
-    void emitValueType(CodeBuilder* builder);
     void emitKey(CodeBuilder* builder, cstring keyName);
-    void emitAction(CodeBuilder* builder, cstring valueName);
-    void emitInitializer(CodeBuilder* builder);
+
+    virtual void emitTypes(CodeBuilder* builder);
+    virtual void emitInstance(CodeBuilder* builder);
+    virtual void emitKeyType(CodeBuilder* builder);
+    virtual void emitValueType(CodeBuilder* builder);
+    virtual void emitValueActionIDNames(CodeBuilder* builder);
+    virtual void emitValueStructStructure(CodeBuilder* builder);
+    virtual void emitAction(CodeBuilder* builder, cstring valueName, cstring actionRunVariable);
+    virtual void emitInitializer(CodeBuilder* builder);
+    virtual void emitLookup(CodeBuilder* builder, cstring key, cstring value) {
+        builder->target->emitTableLookup(builder, dataMapName, key, value);
+        builder->endOfStatement(true);
+    }
+    virtual void emitLookupDefault(CodeBuilder* builder, cstring key, cstring value) {
+        builder->target->emitTableLookup(builder, defaultActionMapName, key, value);
+        builder->endOfStatement(true);
+    }
+    virtual bool isMatchTypeSupported(const IR::Declaration_ID* matchType) {
+        return matchType->name.name == P4::P4CoreLibrary::instance.exactMatch.name ||
+               matchType->name.name == P4::P4CoreLibrary::instance.lpmMatch.name;
+    }
+    // Whether to drop packet if no match entry found.
```

Comment:
this is not really standard P4 semantics.
In P4 a table apply can never cause a packet to be dropped; you have to code that up explicitly.

Reply:
Shouldn't this be 'false' by default?

Reply:
We are fully aware of what the standard P4 semantics is, but the reason for this is how it has been implemented for eBPF model, see the logic before our modifications to this piece of code:

https://github.com/p4lang/p4c/blob/b9a2dcbc34f002cb6b6819dcc543f4caa5e4018a/backends/ebpf/ebpfTable.cpp#L345 

Do you want to change this? If so, this method isn't necessary anymore.

Reply:
So you are saying that the original implementation was already wrong? That's not good. I wonder why I did that.
What does ubpf do?
I think it would be good to fix it.
Perhaps you can do this by having some compiler options that need to be set to trigger the non-standard semantics.
This just in case users depend on this behavior.

Reply:
ubpf does the same, I'm also not sure why.. perhaps we just followed the implementation for eBPF. I agree that it would be good to fix it. In the PSA impl for eBPF we override this and PSA tables don't drop if match isn't found. 

Should I fix it in a separate PR or this one?

Reply:
you can file an issue for it for now so we can keep track.

Reply:
I've opened #3130.